### PR TITLE
Trace Module Implementation

### PR DESCRIPTION
**Pull request by ajlopez**
Including two JSON RPC methods: trace_transaction and trace_block

*fed:newtracetx*

### CODE REVIEW

File: rskj-core/src/main/java/co/rsk/rpc/modules/trace/TraceType.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package co.rsk.rpc.modules.trace;
+
+public enum TraceType {
+    CALL,
+    CREATE,
+    SUICIDE,
+    REWARD
```
[the remark this answers is not in the document]
Comment (ajlopez) [reply rskj-core/src/main/java/co/rsk/rpc/modules/trace/TraceType.java]: REWARD? No, it's not being used. It's in the Parity specification.

File: rskj-core/src/main/java/co/rsk/rpc/modules/trace/TraceAddress.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    public int[] toAddress() {
+        if (this.parent == null) {
+            return EMPTY_ADDRESS;
+        }
+
+        int[] parentAddress = this.parent.toAddress();
+
+        int[] address = new int[parentAddress.length + 1];
+
+        System.arraycopy(parentAddress, 0, address, 0, parentAddress.length);
```
[the remark this answers is not in the document]
Comment (ajlopez) [reply rskj-core/src/main/java/co/rsk/rpc/modules/trace/TraceAddress.java]: done

File: rskj-core/src/main/java/co/rsk/rpc/modules/trace/ProgramSubtrace.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+public class ProgramSubtrace {
+    private final TraceType traceType;
+    private final CallType callType;
+    private final CreationData creationData;
+    private final InvokeData invokeData;
+    private final ProgramResult programResult;
+    private final List<ProgramSubtrace> subtraces;
+
+    public ProgramSubtrace(InvokeData invokeData) {
```
[the remark this answers is not in the document]
Comment (ajlopez) [reply rskj-core/src/main/java/co/rsk/rpc/modules/trace/ProgramSubtrace.java]: done

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 4]
```diff
@@ -419,6 +424,34 @@ public void registerBtcTransaction(Transaction rskTx, byte[] btcTxSerialized, in
...
+        );
+
+        if (this.subtraces != null) {
+            DataWord from = DataWord.valueOf(PrecompiledContracts.BRIDGE_ADDR.getBytes());
+            DataWord to = DataWord.valueOf(receiver.getBytes());
+            long gas = 0L;
+            DataWord value = DataWord.valueOf(amount.getBytes());
+
+            TransferInvoke invoke = new TransferInvoke(from, to, gas, value);
+            ProgramResult result     = new ProgramResult();
```
[the remark this answers is not in the document]
Comment (ajlopez) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java]: done

**Comment by nacho692** [comment]
Pipeline: run

**Comment by nacho692** [comment]
Pipeline: run

**Comment by nacho692** [comment]
Pipeline: run

**Pull request merged**